src/System.Net.HttpListener/tests/HttpListenerAuthenticationTest.cs

```diff
@@ -384,22 +390,21 @@ public void Realm_SetDisposed_ThrowsObjectDisposedException()
 
         public async Task<HttpResponseMessage> AuthenticationFailure(HttpClient client, HttpStatusCode errorCode)
         {
+            client.Timeout = new TimeSpan(0, 0, 1);
             Task<HttpResponseMessage> clientTask = client.GetAsync(_factory.ListeningUrl);
 
             // The server task will hang forever if it is not cancelled.
             var tokenSource = new CancellationTokenSource();
             Task<HttpListenerContext> serverTask = Task.Run(() => _listener.GetContext(), tokenSource.Token);
-
+            
             // The client task should complete first - the server should send a 401 response.
             Task resultTask = await Task.WhenAny(clientTask, serverTask);
             tokenSource.Cancel();
             if (resultTask == serverTask)
             {
-                await serverTask;
+                await clientTask;
```

Comment:
> Awaiting a task that has finished is always a bug in my mind. The code written above will trigger the compiler to generate an iterator class and function which isn't necessary if we already know the task is done

Huh?  The "iterator" and function you mention is for the async method, not for an individual await.  Adding or removing awaits to a method marked as async doesn't impact that.  And it's perfectly fine to await something that's already finished.  And why do we care about compiler optimizations in a test?

> If it's very wrong we shouldn't propagate the exception. Instead we should crash immediately with Debug.Fail and get a dump to better understand the state of both the client and the server (including native code).

This is a test.  Every time any Assert.whatever fails it's something that's "very wrong", and an "Assert.whatever" failing in a test just generates an exception.  This is simply propagating that failure out to the calling test.  I don't understand what the objection here is.